### PR TITLE
Binary serialization

### PR DESCRIPTION
**Pull request by xymus**
Introducing the binary serialization services! (In use in both WBTW and Tinks!) Also fix a few bugs in the serialization phase...

The serialization format is somewhat similar to BSON. The main differences are the support for instances, the precise instance types (as in `Array[MyClass]`), different data types id because we don't have the same data types and that we don't start off with the length of the stream. But we use a similar grammar, null-terminated lists and objects, as well as both a null-terminated string and a block which support UTF8.

The errors system can still be improved. There is too much noise but one can only look at the first error and it is usually enough.

### CODE REVIEW

File: lib/binary/serialization.nit [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+#        | 0x03 id                # Ref to object
+#        | 0x04 int64             # Int
+#        | 0x05 int8              # Bool (int8 != 0)
+#        | 0x06 int8              # Char
+#        | 0x07 double(64 bits)   # Float
+#        | 0x08 block             # String
+#        | 0x09 block             # NativeString
+#        | 0x0A flat_array;       # Array[nullable Object]
+#
+# block = int64 byte*;
```
Comment (privat) [inline lib/binary/serialization.nit]: int64 and byte? Something does not seem right.

File: tests/sav/nitg-e/test_binary_deserialization_alt1.res [hunk 2]
```diff
@@ -0,0 +1,24 @@
+Runtime error: Aborted (../lib/serialization/serialization.nit:109)
```
Comment (privat) [inline tests/sav/nitg-e/test_binary_deserialization_alt1.res]: Is this expected?

**Comment by privat** [comment]
+1

**Comment by xymus** [comment]
Rerolled to fix tests results and remove the remaining abort.

**Comment by R4PaSs** [comment]
:shipit:

**Pull request merged**